### PR TITLE
fix(vrl): mark set_semantic_meaning impure

### LINKED ISSUE #21887: VRL compiler warns about the unused result of set_semantic_meaning
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

The VRL compiler reports a warning for calls to `set_semantic_meaning`:

```
2024-11-26T02:02:17.453462Z  WARN transform{component_kind="transform" component_id=talos_kernel_logs_cleanup component_type=remap}: vector::transforms::remap: VRL compilation warning. warnings=
warning[E900]: unused result for function call `set_semantic_meaning(."talos-time", "timestamp")`
  ┌─ :1:1
  │
1 │ set_semantic_meaning(."talos-time", "timestamp")
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the result of this expression or remove it
  │
  = this expression has no side-effects
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples
```

From looking at the docs, it seems to be true that the function has no side effects, but that it also returns no value so the warning seems wrong.

### Configuration

```text
set_semantic_meaning(."talos-time", "timestamp")

del(.clock)
del(.seq)
del(."talos-level")
."talos-time" = parse_timestamp!(."talos-time", "%+")
```

### Version

0.42.0

### Debug Output

_No response_

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by sfackler**
## Summary
The VRL compiler previously warned about the unused result of `set_semantic_meaning`, even though it doesn't return anything due to it being incorrectly considered a pure function.

## Change Type
- [ ] Bug fix
- [ ] New feature
- [x] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [x] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- [ ] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

Closes vectordotdev/vrl#1149

### CODE REVIEW

**Review by pront — APPROVED** [review]
Thanks @sfackler! Did you have a chance to test that the warning went away?

**Comment by sfackler** [comment]
I have not - what's the easiest way to do that?

**Comment by sfackler** [comment]
Huh - it seems like this change doesn't actually fix the warning? I had to add a second statement to the VRL script (does it treat the last expression as an implicit return to the caller?) for the warning to trigger at all, but with that setup it warns both on master and this branch:

```
$ cat config.yaml
sources:
  demo_logs:
    type: demo_logs
    format: json
transforms:
  remap:
    type: remap
    inputs: [demo_logs]
    source: |
      set_semantic_meaning(.foo, "a")
      .foo = "bar"
sinks:
  console:
    type: console
    inputs: [remap]
    encoding:
      codec: json
$ cargo run -p vector -- --config config.yaml
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.33s
     Running `target/debug/vector --config config.yaml`
2024-11-26T21:09:00.300855Z  INFO vector::app: Log level is enabled. level="info"
2024-11-26T21:09:00.319805Z  INFO vector::app: Loading configs. paths=["config.yaml"]
2024-11-26T21:09:00.329959Z  WARN transform{component_kind="transform" component_id=remap component_type=remap}: vector::transforms::remap: VRL compilation warning. warnings=
warning[E900]: unused result for function call `set_semantic_meaning(.foo, "a")`
  ┌─ :1:1
  │
1 │ set_semantic_meaning(.foo, "a")
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the result of this expression or remove it
  │
  = this expression has no side-effects
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples

2024-11-26T21:09:00.331415Z  INFO vector::topology::running: Running healthchecks.
2024-11-26T21:09:00.331730Z  INFO vector::topology::builder: Healthcheck passed.
2024-11-26T21:09:00.331916Z  INFO vector: Vector has started. debug="true" version="0.43.0" arch="x86_64" revision=""
2024-11-26T21:09:00.332007Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
{"foo":"bar"}
```

**Comment by sfackler** [comment]
Oh - looks like the lint doesn't actually check if the function is pure or not: https://github.com/vectordotdev/vrl/blob/e3852056d7f2f52a6127c5407ab840f52e545a88/src/compiler/unused_expression_checker.rs#L354-L372

**Comment by pront** [comment]
> does it treat the last expression as an implicit return to the caller?

Yes, the result of the last expression is considered "used". I think we need to add this function to this [list](https://github.com/vectordotdev/vrl/blob/main/src/compiler/unused_expression_checker.rs#L356). 

But we should also merge this PR, since this a 'strange' function which modifies internal state.

File: changelog.d/21896_set_semantic_meaning_impure.fix.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+The `set_semantic_meaning` VRL function is now marked impure to avoid incorrectly triggering the
```
Comment (pront) [inline changelog.d/21896_set_semantic_meaning_impure.fix.md:1]: We should remove this changelog fragment

File: config.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+sources:
```
Comment (pront) [inline config.yaml:1]: Should also `git rm` this one. Probably was committed by accident.

Comment (sfackler) [reply config.yaml]: Oops, fixed

**Review by pront — APPROVED** [review]

**Pull request merged**